### ISSUE TITLE
Where is soft relu from?

### ISSUE THREAD

**Issue by kruus** [oneapi-src/oneDNN#728]
On VE:
I see benchdnn failing backward srelu test every time input is -90 (other vals OK)
`benchdnn --eltwise -v99 --dir=BWD_D --dt=f32 --tag=aBx8b --alg=srelu 4x27x2x2x2`
perhaps also affecting the gtest

Will compare with x86.

**Comment by kruus** [comment]
oops log1pf is ln(1+x), whew.

**Comment by vpirogov** [comment]
All the formulas for eltwise are [here](https://oneapi-src.github.io/oneDNN/dev_guide_eltwise.html), in case you are interested in anything else.

**Comment by kruus** [comment]
Thanks @vpirogov .   benchdnn docs for `--alg` refer to a similar table, but it wasn't until I looked
at `tests/benchdnn/inputs/eltwise/test_eltwise_all` that I figured out that

table has **soft_relu**; but benchdnn --eltwise wants --alg= **srelu**,
Hmmm. Were the `_dst` suffixes, like `--alg=relu_dst`, documented? Don't recall atm.

My real issue turned out to be benchdnn very low input value `-90` with nc++.

Scalar loops using stock `::expf` seem to generate zero nicely for `soft_relu_bwd`.
However, when nc++ takes vector reciprocal of `1 + __vec_exp` I can get **nans**.
A consequene of agressive vectorization, it seems.

After a few options, I'm considering the following, which vectorizes decently using masked ops:
```
template <typename T, typename U = typename utils::remove_reference<T>::type>
inline U soft_relu_bwd(T dd, T s) {
#if defined(__ve)
    return s <= U{-87} ? U{0} : (U)(dd / (1 + ::expf((float)(-s))));
#else
    return (U)(dd / (1 + ::expf((float)(-s))));
#endif
}
```

I may need similar checks for `swish_fwd`,  `logistic_fwd` etc. which are similar functions.

**Comment by vpirogov** [comment]
Adding @dzarukin and @mgouicem to comment on benchdnn behavior and numerical question.

**Comment by dzarukin** [comment]
Hi @kruus,
I see multiple issues raised.

Regarding the main one with nans. It looks like the problem is reciprocal but not the math itself. It most likely (I haven't checked yet, but I will) that reciprocal of `0` or `inf` values is `NaN` but if regular fp division is applied, result will be as expected: `1/0 = inf` and `1/inf = 0`. If there's a limitation for nc++ to apply fp division, we may patch the library to be reciprocal friendly.

Regarding missing `_dst` algs in benchdnn documentation - yes, they are missing, it will be added to master soon. `_dst` are only mentioned in dev_guide for now.

Regarding mismatch of alg names - yes, some alg names mismatch in the library and in benchdnn. That was done to shorten the command-line inputs. We will consider adding aliases in benchdnn for algorithms to support alg names without `dnnl_eltwise_` as they states in [dev_guide](https://oneapi-src.github.io/oneDNN/dev_guide_eltwise.html) (not for _use_dst_for_bwd though).

Will all of this work for you?

**Comment by kruus** [comment]
Yes, more than satisfactory, seeing as I closed the question some days ago :)  No hurry either, as I've a number of other issues to track down :).  This may well only be an issue with VRCP op on VE chip.

So if x86 is immune to vector reciprocal issues, feel free to keep master "clean" -- I now put my patches in src/XXX/ve/ and the cmake build sources [for nc++] simply replace any "same-name" files in `src/XXX/` with my hacked `ve/` versions.

**Comment by dzarukin** [comment]
@kruus, just to confirm, how do you expect to treat `log_bwd = 1 / x`? Does nc++ return `INF` or `NaN` as well? Would you expect to get `INF` every time for x = 0? Same question about sqrt_bwd.

**Comment by kruus** [comment]
- Wrote a little test program covering many limit cases on VE
- test includes `numeric_limits<float>` special values, zero, and a selection covering the range of +ve numeric values
- Here's the discrepancies I saw for scalar vs vectorized `1/x` on VE:

| x | scalar 1/x | VRCP 1/x |
| --- | --- | --- |
| inf | 0 | -nan |
| -inf | -0 | -nan |
| denorm_min() | inf | -nan |
| 0 | inf | -nan |
| **no diff for:** | --- | --- |
| quiet_NaN() | -nan | same |
| signaling_NaN() | -nan | same |

For non-zero numeric `x`, scalar and vector `1/x` values seem to agree.

Is x86 VRCPSS behavior more sane?

**Comment by rsdubtso** [comment]
Code:
```c++
#include <stdio.h>
#include <limits>

// This assumes Unix ABI and may or may not work :)
void __attribute__ ((noinline)) foo(float *x) {
    asm volatile (
            "vmovups (%rdi), %ymm0\n\t"
            "vrcpps %ymm0, %ymm0\n\t"
            "vmovups %ymm0, (%rdi)\n\t"
            );
}

int main(int argc, char **argv) {
    float x[8] = {
        +std::numeric_limits<float>::infinity(),
        -std::numeric_limits<float>::infinity(),
        std::numeric_limits<float>::denorm_min(),
        0,
        std::numeric_limits<float>::quiet_NaN(),
        std::numeric_limits<float>::signaling_NaN()
    };
    printf("x = {%g %g %g %g %g %g}\n",
            x[0], x[1], x[2], x[3], x[4], x[5]);
    foo(x);
    printf("vrcpps(x) = {%g %g %g %g %g %g}\n",
            x[0], x[1], x[2], x[3], x[4], x[5]);
    return 0;
}
```
Output:
```
x = {inf -inf 1.4013e-45 0 nan nan}
vrcpps(x) = {0 -0 inf inf nan nan}
```

The `vrcp14ps` instruction behaves the same way.

**Comment by kruus** [comment]
nc++ vectorization with sqrt and 1/sqrt is a mess for limit cases too.  Only the scalar function calls behaved sanely. Why did you ask about it !? I wish I had never looked.

Please restrict your attention only to documentation issues and close.

I'll slowly fix (on fork) only stuff that causes actual test failures -- there's far too much fixing to do it properly right now.

#### nc++ sqrt notes:

- On VE scalar sqrt(x) gives `{inf -nan 0 0 -nan -nan}`
  - `__fast_sqrtf` is reasonable
- but vectorized `sqrt(x)` gave `{-nan -nan 0 -nan -nan}` . Surprise `sqrt(+inf)=-nan`
  - Surprisingly, nc++ vectorized sqrt(x) as `double(x) * (1/sqrt((double)x)`, two iterative fixups, and a final convert back to float.  So instead seeing vector float sqrt `vfsqrt.f`, I see a `vrsqrt.d.nex` op!

- On VE scalar 0.5/sqrt(x) gives `{0 -nan inf inf -nan -nan}`
  - uses __fast_sqrtf, fmul, fdiv
- vector 0.5/sqrt(x) gave `{-nan -nan -nan -nan -nan}`
  - using ... well ... absolutely **braindead** assembler doing something like calc `x*(1/sqrt((double)x)`, with iterative correstions, then a vector reciprocal with iterative corrections, then finally back to float.
  - Even adapting [the Quake method](https://stackoverflow.com/questions/11644441/fast-inverse-square-root-on-x64) `f[i] = 0.5*inv_sqrt<float,2>(x[i])` vectorized better, although still with some strangeness for limit cases `{inf inf 1.48633e+19 1.48633e+19 -nan -nan}`:
```
template <typename T, char iterations = 2> inline T inv_sqrt(T x) {
    static_assert(std::is_floating_point<T>::value, "T must be floating point");
    static_assert(iterations == 1 or iterations == 2, "itarations must equal 1 or 2");
    typedef typename std::conditional<sizeof(T) == 8, std::int64_t, std::int32_t>::type Tint;
    T y = x;
    T x2 = y * T{0.5};
    Tint i = *(Tint *)&y;
    i = (sizeof(T) == 8 ? 0x5fe6eb50c7b537a9 : 0x5f3759df) - (i >> 1);
    y = *(T *)&i;
    y = y * (T{1.5} - (x2 * y * y));
    if (iterations == 2)
        y = y * (T{1.5} - (x2 * y * y));
    return y;
}
```

**Comment by vpirogov** [comment]
Resolved by 5699ec3cae4fa0c444410980a34630cd4797dc80 and 07e97e40683b42ec53cd5835f4086be8616794c7.
